torch/nn/parameter.py

```diff
@@ -44,3 +44,93 @@ def __reduce_ex__(self, proto):
         )
 
     __torch_function__ = _disabled_torch_function_impl
+
+
+class UninitializedParameter(Parameter):
+    r"""A parameter that can be lazily initialized.
+
+    Uninitialized Parameters holds empty tensors that can be moved
+    to different devices and change their type so when they are materialized,
+    they do it directly with the right configuration. Moreover, Uninitialized parameters
+    become full fledged :class:`torch.nn.Parameter` after they are initialized.
+    """
+    _allowed_methods = [
+        torch.Tensor.__hash__,
```

Comment:
What if dunder hash, size, and copy_ weren't supported?

Reply:
These are needed to support manipulating the `UninitializedParameters` for the materialization and loading a state_dict. I think this needs to be documented.

Reply:
Are they still needed if state dicts are unsupported? Also, if "materialization" constructs new parameters then I don't think we need these, either?

Reply:
materialize currently changes the parameter data in place. 
Also loading a state dict should be supported and this triggers a `copy_` call,'
`__hash__` is used when iterating through the module parameters, so it's actually being called in a bunch of different places.

Reply:
OK, sounds good for now.